### PR TITLE
Succesful Edge Movement

### PR DESCRIPTION
**Pull request by wfletch**
Some changes:

Nodes now tick EITHER pre-load edges or inbound edges but not both
Once a car reaches it's destination it is popped and then just... forgotten about.
Added logic to make sure we remove a car from one edge before moving to another.
Copious amounts of debugging print statements

### CODE REVIEW

File: simulation_structure.py [hunk 1]
```diff
@@ -27,9 +28,10 @@ def output_network_state(self):
         print(network_dump)
 
         
-    def tick(self):     
+    def tick(self):
+        print("TICK:", self.tick_count)    
         self.nm.tick()
-        self.output_network_state()
```
Comment (wfletch) [inline simulation_structure.py:32]: Removed as it is just adding noise right now.

File: simulation_structure.py [hunk 2]
```diff
@@ -66,11 +68,15 @@ def place_new_car(self, car_entry): #None
...
+        edge_list = list(self.edge_id_to_edge_mapping.keys())   
         random.shuffle(node_list)   # ensure no node nor edge's inbound traffic favoured
         for node_id in node_list:
-            print(node_id)
             node = self.node_id_to_node_mapping[node_id]
             node.node_tick()
+        for edge_id in edge_list:
+            edge = self.edge_id_to_edge_mapping[edge_id]
+            edge.shift_cars_up()
+            print(edge.id, edge.queue)
```
Comment (wfletch) [inline simulation_structure.py:79]: So, I think we need to tick on each edge.

the only thing an edge tick does is increase the position by one and add any new cars to the queue.
In future we would process this AFTER the corresponding NODE has processed the edge. But for now it is

Step 1: Process Nodes
Step 2: Process Edges
Step 3: Profit

File: simulation_structure.py [hunk 3]
```diff
@@ -103,39 +109,47 @@ def get_car_to_move(self, inbound_edge):
...
+                    # exception needed:  car cannot move as intended, see if recalculation possible
+                    raise Exception("that edge does not exist")
+                elif next_edge.has_space_for_new_car():
+                    print("We have space for a new car")
+                    if not next_edge.pre_loaded_cars[0]:  # if no cars in wait -- TODO:  set up function in car,
+                        print("We are adding a car!")
+                        next_edge.pre_loaded_cars.append(car_to_add)
+                        print ("NEXT EDGE QUEUE:", next_edge.pre_loaded_cars)
+                        # next_edge.pre_loaded_cars.append(None)   # clear waiting queue
+                return
```
Comment (wfletch) [inline simulation_structure.py:125]: This return statement ensures we only process the pre-loaded cars OR new inbound.

File: simulation_structure.py [hunk 4]
```diff
@@ -103,39 +109,47 @@ def get_car_to_move(self, inbound_edge):
...
-                    # exception needed:  car cannot move as intended, see if recalculation possible
-                    raise Exception("that edge does not exist")
-                elif next_edge.has_space_for_new_car():
-                    if not next_edge.pre_loaded_cars[0]:  # if no cars in wait -- TODO:  set up function in car,
-                        next_edge.pre_loaded_cars.append(car_to_move)
-                        inbound_edge.shift_cars_up()
-                        next_edge.pre_loaded_cars.append(None)   # clear waiting queue
-
-                # TODO: add +1 to car_on_network_time ==> use to calculate delays ==> VERSION 2
+                if next_edge_id != None:
```
Comment (wfletch) [inline simulation_structure.py:138]: next_edge_id could be 0 which would be a valid number, but if 0 is false.

FUN FACT

**Review by wfletch — COMMENTED** [review]
Explanations of all changes.

File: simulation_structure.py (hunk 2 above)
Comment (julialruiter) [reply simulation_structure.py:79]: this is far more elloquent and consistent with node logic than the workaround I was working on

File: simulation_structure.py (hunk 3 above)
Comment (julialruiter) [reply simulation_structure.py:125]: I forgot the else before, meaning both were occurring

File: simulation_structure.py (hunk 4 above)
Comment (julialruiter) [reply simulation_structure.py:138]: this wasn't a true/false though?  It was already list/none?

**Pull request merged**